File: src/co/nums/intellij/aem/htl/inspections/HtlStringQuotesFix.java
package co.nums.intellij.aem.htl.inspections;

import co.nums.intellij.aem.htl.psi.HtlTokenTypes;
import com.intellij.codeInsight.FileModificationService;
import com.intellij.codeInsight.intention.IntentionAction;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

class HtlStringQuotesFix implements IntentionAction {

	private static final char DOUBLE_QUOTE = '"';
	private static final char SINGLE_QUOTE = '\'';
	private static final String NAME = "Fix quotes";

	private final char currentQuote;
	private final char quoteToInsert;

	HtlStringQuotesFix(char currentQuote) {
		this.currentQuote = currentQuote;
		this.quoteToInsert = currentQuote == DOUBLE_QUOTE ? SINGLE_QUOTE : DOUBLE_QUOTE;
	}

	@Nls
	@NotNull
	@Override
	public String getText() {
		return NAME;
	}

	@Nls
	@NotNull
	@Override
	public String getFamilyName() {
		return NAME;
	}

	@Override
	public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
		return true;
	}

	@Override
	public boolean startInWriteAction() {
		return true;
	}

	@Override
	public void invoke(@NotNull Project project, Editor editor, PsiFile file) {
		PsiElement element = getCurrentElement(editor, file);
		if (isHtlString(element) && editionAllowed(element)) {
			fixQuotes(project, file, element);
		}
	}

	private static PsiElement getCurrentElement(Editor editor, PsiFile file) {
		int offset = editor.getCaretModel().getOffset();
		return file.findElementAt(offset);
	}

	private static boolean isHtlString(PsiElement element) {
		if (element instanceof LeafPsiElement) {
			LeafPsiElement leafElement = (LeafPsiElement) element;
			return leafElement.getElementType() == HtlTokenTypes.SINGLE_QUOTED_STRING
					|| leafElement.getElementType() == HtlTokenTypes.DOUBLE_QUOTED_STRING;
		}
		return false;
	}

	private static boolean editionAllowed(PsiElement element) {
		return element != null && element.isValid()
				&& FileModificationService.getInstance().prepareFileForWrite(element.getContainingFile());
	}

	private void fixQuotes(@NotNull Project project, PsiFile file, PsiElement element) {
		Document document = PsiDocumentManager.getInstance(project).getDocument(file);
		if (document != null) {
			String currentText = element.getText();
			String replacement = quoteToInsert + currentText.substring(1);
			if (endsWithUnescapedQuote(currentText, replacement)) {
				replacement = replacement.substring(0, replacement.length() - 1) + quoteToInsert;
			}
			TextRange textRange = element.getTextRange();
			document.replaceString(textRange.getStartOffset(), textRange.getEndOffset(), replacement);
		}
	}

	private boolean endsWithUnescapedQuote(String currentText, String replacement) {
		return currentText.length() > 1
				&& currentText.endsWith(Character.toString(currentQuote)) && !replacement.endsWith("\\" + currentQuote);
	}

}

File: src/co/nums/intellij/aem/htl/inspections/HtlStringQuotesFix.kt
package co.nums.intellij.aem.htl.inspections

import co.nums.intellij.aem.htl.psi.HtlTokenTypes
import com.intellij.codeInsight.FileModificationService
import com.intellij.codeInsight.intention.IntentionAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiDocumentManager
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.impl.source.tree.LeafPsiElement

private const val NAME = "Fix quotes"
private const val SINGLE_QUOTE = '\''
private const val DOUBLE_QUOTE = '"'

object HtlStringSingleQuotesFix : HtlStringQuotesFix(SINGLE_QUOTE)
object HtlStringDoubleQuotesFix : HtlStringQuotesFix(DOUBLE_QUOTE)

open class HtlStringQuotesFix(private val quoteToFix: Char) : IntentionAction {

    private val quoteToInsert: Char
        get() = if (quoteToFix == DOUBLE_QUOTE) SINGLE_QUOTE else DOUBLE_QUOTE

    override fun invoke(project: Project, editor: Editor, file: PsiFile) {
        val currentElement = file.findElementAt(editor.caretModel.offset) ?: return
        if (currentElement.isHtlString() && currentElement.canBeEdited()) {
            fixQuotes(project, file, currentElement)
        }
    }

    private fun PsiElement.isHtlString() =
            this is LeafPsiElement
                    && (this.elementType === HtlTokenTypes.SINGLE_QUOTED_STRING
                    || this.elementType === HtlTokenTypes.DOUBLE_QUOTED_STRING)

    private fun PsiElement.canBeEdited() =
            this.isValid && FileModificationService.getInstance().prepareFileForWrite(this.containingFile)

    private fun fixQuotes(project: Project, file: PsiFile, htlStringLiteral: PsiElement) {
        val document = PsiDocumentManager.getInstance(project).getDocument(file)
        if (document != null) {
            val currentText = htlStringLiteral.text
            var replacement = quoteToInsert + currentText.substring(1)
            if (endsWithUnescapedQuote(currentText, replacement)) {
                replacement = replacement.substring(0, replacement.length - 1) + quoteToInsert
            }
            val textRange = htlStringLiteral.textRange
            document.replaceString(textRange.startOffset, textRange.endOffset, replacement)
        }
    }

    private fun endsWithUnescapedQuote(currentText: String, replacement: String) =
            currentText.length > 1
                    && currentText.endsWith(Character.toString(quoteToFix)) && !replacement.endsWith("\\" + quoteToFix)

    override fun getText() = NAME

    override fun getFamilyName() = NAME

    override fun startInWriteAction() = true

    override fun isAvailable(project: Project, editor: Editor, file: PsiFile) = true

}
